src/ke.cpp

```diff
@@ -124,24 +193,44 @@ KeQueryMaximumProcessorCountEx(_In_ USHORT group_number) { return GetMaximumProc
 KAFFINITY
 KeSetSystemAffinityThreadEx(KAFFINITY affinity)
 {
-    uintptr_t old_affinity = SetThreadAffinityMask(GetCurrentThread(), affinity);
-    if (old_affinity == 0) {
+    GROUP_AFFINITY old_affinity;
+    usersim_get_current_thread_group_affinity(&old_affinity);
+    _usersim_thread_affinity.Group = old_affinity.Group;
+    _usersim_thread_affinity.Mask = affinity;
+    if (KeGetCurrentIrql() < DISPATCH_LEVEL && SetThreadAffinityMask(GetCurrentThread(), affinity) == 0) {
         unsigned long error = GetLastError();
         KeBugCheckEx(0, error, 0, 0, 0);
     }
-    return (KAFFINITY)old_affinity;
+    return (KAFFINITY)old_affinity.Mask;
 }
 
 _IRQL_requires_min_(PASSIVE_LEVEL) _IRQL_requires_max_(APC_LEVEL) NTKERNELAPI VOID
     KeRevertToUserAffinityThreadEx(_In_ KAFFINITY affinity)
 {
-    SetThreadAffinityMask(GetCurrentThread(), affinity);
+    KeSetSystemAffinityThreadEx(affinity);
 }
 
 PKTHREAD
 NTAPI
 KeGetCurrentThread(VOID) { return (PKTHREAD)usersim_get_current_thread_id(); }
 
+void
+usersim_get_current_thread_group_affinity(_Out_ GROUP_AFFINITY* Affinity)
```

Comment:
```suggestion
usersim_get_current_thread_group_affinity(_Out_ GROUP_AFFINITY* affinity)
```
FYI, use lower_snake_case in this project.  Yeah, it wasn't consistent before, and I'm cleaning it up :)